src/core/observables/ProfileObservable.hpp

```diff
@@ -33,13 +36,40 @@ class ProfileObservable : virtual public Observable {
         max_z(max_z), n_x_bins(static_cast<size_t>(n_x_bins)),
         n_y_bins(static_cast<size_t>(n_y_bins)),
         n_z_bins(static_cast<size_t>(n_z_bins)) {}
+  // Range of the profile edges.
   double min_x, max_x;
   double min_y, max_y;
   double min_z, max_z;
+  // Number of bins for each coordinate.
   size_t n_x_bins, n_y_bins, n_z_bins;
+
   std::vector<size_t> shape() const override {
     return {n_x_bins, n_y_bins, n_z_bins};
   }
+
+  /** Calculate the bin edges for each dimension */
+  std::vector<std::vector<double>> edges() {
```

Comment:
This whole class seems to be identical `CylindricalProfileObservable` except for the variable names, maybe this repetition can be avoided?

Reply:
I spent 10 days refactoring these observables classes (#3599), but the `AutoParameters` framework keeps blocking me no matter what I try. I gave up using class templates, otherwise I won't get anything done.

Reply:
But this is the core code, no? Also I don't think even a template is needed, they can just be the sample class, the only difference is the name of the data members for the size and the limits?

```c++
class ProfileObservable : virtual public Observable {
 std::array<std::pair<double, double>, 3> m_limits;
std::array<size_t, 3> m_bins;
public:
   ProfileObservable = ...

  std::vector<size_t> shape() const override {
    return {m_bins.begin(), m_bins.end()};
  }

  std::array<std::vector<double>, 3> edges() {
   std::array<std::vector<double>, 3> ret;

  for(i in {0, 1, 2}) {
   boost::copy(Utils::make_lin_space(m_limits[i].first, m_limits[i].second, bins[i]),     std::back_inserter(profile_edges[i]));
   } 
    return ret;
  }
};

class CartesianProfileObservable : public ProfileObservable {
  public:
  size_t n_bins_x() const { return m_bins[0]; }
 ... 
}

class CylindricalProfileObservable : public ProfileObservable {
  public:
  size_t n_bins_phi() const { return m_bins[0]; }
 ... 
}
```
I'm not even sure if the derived classes are really needed. If not
you can use the same interface. Otherwise you don't have to use
the fact that they have a common base in the script interface, so
it should not matter there. Disclaimer: Just for illustration, but I 
*think* something like this shoudl work?

Reply:
I had another solution in mind, with a template class that takes a coordinate system struct as template parameter, and where the `n_bins_*`... are aliases (e.g. `size_t &n_bins_x = n_bins_0`) to avoid writing getters. I would prefer to defer this change to another PR to keep this one simple, and because such a change would have to be carried out on the `LBProfileObservable` classes as well, which will then collide with #3607.

Reply:
If you don't want to change this now, you cold also put the implementation of `edges()` into a free function an reuse it without messing with the class hierarchy.

Reply:
I've attempted it in [jngrad/espresso:observable-edges-4](https://github.com/jngrad/espresso/tree/observable-edges-4). `AutoParameters` complains about implicitly deleted ctors (5cc44a82cb). Making `ProfileObservable` a stand-alone class instead of inheriting from `Observable` solves the issue (c3c9afca978). I couldn't reproduce the compiler error with a toy example. What am I doing wrong in 5cc44a82cb?

Reply:
If you add reference data members, the class is no longer default constructible, because reference can not be unbound. Just use functions, your construction looks fishy anyway.

Reply:
I've converted them to getters in [jngrad/espresso:observable-edges-5](https://github.com/jngrad/espresso/tree/observable-edges-5) but I still get the same compiler errors, plus new ones that are so long they don't even fit on my fullscreen terminal window. Probably caused by a typo, but I don't have the patience to dissect these 18136-character longer compiler warnings.

Reply:
This is interesting. From what I can tell the problem is already with the core classes, e.g. `Observables::DensityProfile` can not be constructed. This seems to be related to virtual
inheritance, or the lack thereof to be more precise. The following very simple unit test
shows the issue:

```c++
#define BOOST_TEST_MODULE Observables::DensityProfile
#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>

#include "observables/DensityProfile.hpp"

BOOST_AUTO_TEST_CASE(ctor) {
  Observables::DensityProfile densityProfile({}, {}, {}, {}, {}, {}, {}, {}, {}, {});
}
```

Leading to
```
[...]
/ssd/fweik/espresso/src/core/observables/DensityProfile.hpp:33:31: error: no matching function for call to ‘Observables::ProfileObservableBase::ProfileObservableBase()’
```
which lets me suspect that there is a second base `ProfileObservableBase`, in addition to the one whose ctor is explicitly called from `ProfileObservableBase`, which is then tried to default
construct, which fails, because there is no default ctor.

I don't understand how this happens, nor why there is `virtual` inheritance in this hierarchy in the first place?!

Reply:
At least for `DensityProfile`, the issue can be fixed by not deriving `ProfileBase` from `Observable`, and implementing `shape` in the derived classes (most of them by either just calling `ProfileBase::shape`, or adding something to the result).

Reply:
> the issue can be fixed by not deriving `ProfileBase` from `Observable`

This is also what I ended up doing in c3c9afc.